### ISSUE TITLE
Handling intermittent data retrieval errors

### ISSUE THREAD

**Issue by RichardScottOZ** [gjoseph92/stackstac#18]
Now and then it seems a download fails - not because it doesn't exist, just one of those internet things.  Even with excellent AWS data to AWS s3.

![image](https://user-images.githubusercontent.com/72196131/112733689-5be9a200-8f91-11eb-817b-17fa245d821f.png)

This is the usual error:

CPLE_OpenFailedError: '/vsicurl/https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/22/M/EV/2020/7/S2A_22MEV_20200704_0_L2A/B02.tif' not recognized as a supported file format.

So something where it can retry when it is not a 404?

**Comment by gjoseph92** [comment]
Agreed, adding some retry logic could be appropriate, both to dataset opens and reads. Though it might be hard to identify which errors are appropriate to retry. "not recognized as a supported file format" sure doesn't sound like something that you should retry.

We'd need to look through the vsicurl -> GDAL -> rasterio logic a bit to understand how HTTP error codes map onto the Python error that's ultimately raised.

**Comment by RichardScottOZ** [comment]
Yes, I have seen that 'not recognised as a file format' multiple times for data that is absolutely there. e.g. try it again the next time, perfectly fine.

I downloaded that whole file by hand and viewed it no problem, just to double check - so it would seem to be some sort of didn't read/network issue meaning it fails to get the identification correctly - your point on the GDAL rasterio errors.

**Comment by RichardScottOZ** [comment]
Doing some more testing...same behaviour, 404 when my internet problem... vsis error when element84 missing data

**Comment by gjoseph92** [comment]
@TomAugspurger pointed out in https://github.com/microsoft/PlanetaryComputer/issues/11#issuecomment-995380986 that we could use `GDAL_HTTP_MAX_RETRY` and `GDAL_HTTP_RETRY_DELAY`, which would be very easy to add to the default GDAL options: https://github.com/gjoseph92/stackstac/blob/d3a78c490860c85319a91bdf3712e58986593def/stackstac/rio_reader.py#L36-L42

**Comment by charalamm** [comment]
@gjoseph92 unfortunately [GDAL_HTTP_MAX_RETRY](https://gdal.org/user/virtual_file_systems.html#:~:text=Starting%20with%20GDAL%202.3%2C%20the%20GDAL_HTTP_MAX_RETRY%20(number%20of%20attempts)%20and%20GDAL_HTTP_RETRY_DELAY%20(in%20seconds)%20configuration%20option%20can%20be%20set%2C%20so%20that%20request%20retries%20are%20done%20in%20case%20of%20HTTP%20errors%20429%2C%20502%2C%20503%20or%20504.) helps only for retrying HTTP errors 429, 502, 503 or 504. As also said in other comments, I also think it would be nice to retrying support user-specific errors

**Comment by charalamm** [comment]
@gjoseph92 I am trying to solve this with the approach mentioned above. For this I tried to separate the dataset reader creation in a different function so that users can overwrite it with a retry. My changes in stackstac can be found [here](https://github.com/gjoseph92/stackstac/compare/main...charalamm:stackstac:retry-user-difined-errors?expand=1).

Later I tried to overwrite the method like:
```
import stackstac
from rasterio import RasterioIOError
from time import sleep

class AutoParallelRioReaderWithRetry(stackstac.rio_reader.AutoParallelRioReader):
    def _get_ds(self) -> stackstac.rio_reader.SelfCleaningDatasetReader:
        """
        Retry
        """
        retries = 10
        retries_delay = 10

        for _ in range(retries):
            try:
                return stackstac.rio_reader.SelfCleaningDatasetReader(self.url, sharing=False)
            except RasterioIOError as ex:
                exception = ex
                dns_problem_condition = ("Could not resolve host" in str(ex))
                read_problem_condition = ("not recognized as a supported dataset name" in str(ex))
                if dns_problem_condition or read_problem_condition:
                    print("retrying")
                    sleep(retries_delay)
                    continue
                print(f"Failed to open {self.url} with exception {ex}")
                raise ex
        raise Exception(f"Failed to open {self.url} after {retries} retries with error {exception}")
```

However I notice that the same readers are failing again and again within the same compute run although they succeed in different runs. I even had cases when compute run without any problem. Is there any place that the result of SelfCleaningDatasetReader is getting cached?
